lambda/matching.ml

```diff
@@ -210,73 +269,82 @@ end = struct
     | _ -> p
 
   let of_clause ~arg cl =
-    let rec aux ((pat, patl), action) =
-      match pat.pat_desc with
-      | Tpat_any -> ((pat, patl), action)
-      | Tpat_var (id, s) ->
-          let p = { pat with pat_desc = Tpat_alias (omega, id, s) } in
-          aux ((p, patl), action)
-      | Tpat_alias (p, id, _) ->
-          let k = Typeopt.value_kind pat.pat_env pat.pat_type in
-          aux ((p, patl), bind_with_value_kind Alias (id, k) arg action)
-      | Tpat_record ([], _) -> ((omega, patl), action)
-      | Tpat_record (lbls, closed) ->
-          let all_lbls = all_record_args lbls in
-          let full_pat =
-            { pat with pat_desc = Tpat_record (all_lbls, closed) }
-          in
-          ((full_pat, patl), action)
-      | Tpat_or _ -> (
-          let pat_simple = simpl_orpat pat in
-          match pat_simple.pat_desc with
-          | Tpat_or _ -> ((pat_simple, patl), action)
-          | _ -> aux ((pat_simple, patl), action)
+    let rec aux (((p, patl), action) : General.clause) : clause =
+      let continue p (view : general_view) : clause =
+        aux (({ p with pat_desc = view }, patl), action)
+      in
+      let stop p (view : half_simple_view) : clause =
+        (({ p with pat_desc = view }, patl), action)
+      in
+      match p.pat_desc with
+      | `Any -> stop p `Any
+      | `Var (id, s) -> continue p (`Alias (omega, id, s))
+      | `Alias (p, id, _) ->
+          let k = Typeopt.value_kind p.pat_env p.pat_type in
+          aux
+            ( (General.view p, patl),
+              bind_with_value_kind Alias (id, k) arg action )
+      | `Record ([], _) as view -> stop p view
```

Comment:
The head of this pattern changed from `` `Any `` to `` `Record ``. Since this case cannot happen in the surface language, this changes is undetectable. However, I would prefer to keep the previous behavior to make sure that all records are extended in half compiler pattern are fully extended (it looks like extending them is not an option since `all_record_args` requires on label to compute the list of all labels).

Reply:
Maybe we could try to expand record labels from the `pat_type` rather than a label? (Using `Any` here has the property that you give, but it is also wrong semantically, at least in an untyped model.)

Reply:
I just remarked that the `simplify_under_or` function raises a fatal_error on empty record patterns. So it definitely does not make sense to spend too much type decorating this specific empty set.